src/main.rs

```diff
@@ -127,38 +132,83 @@ impl CargoCommand {
             _ => (command, false),
         };
 
+        let message_format = match Self::extract_message_format(&mut remaining_args) {
+            Some(format) => {
+                if !format.starts_with("json") {
+                    eprintln!("error: non-JSON `message-format` is not supported");
+                    std::process::exit(1);
+                }
+                format
+            }
+            None => Self::DEFAULT_MESSAGE_FORMAT.to_string(),
+        };
+
         Some(Self {
             command,
             should_link,
             args: remaining_args,
+            message_format,
         })
     }
 
+    fn extract_message_format(args: &mut Vec<String>) -> Option<String> {
+        for (i, arg) in args.iter().enumerate() {
+            if arg.starts_with("--message-format") {
+                return {
+                    let arg = args.remove(i);
```

Comment:
Wow, rustc understands that the immutable borrow from `iter` is no longer needed at this point due to the return and not referencing `arg`, and lets you call `remove` (mutable borrow). Sweet!

Reply:
Yeah, I wasn't really expecting this to compile when I first wrote it, but was impressed that that it actually did!